lib/guard_against_physical_delete/support_counter_cache/associations/builder/belongs_to.rb

```diff
@@ -27,29 +24,56 @@ def belongs_to_counter_cache_after_logical_delete(reflection)
                   def belongs_to_counter_cache_after_revive(reflection)
                     cache_column = reflection.counter_cache_column
                     delete_column = self.logical_delete_column
-                    return unless send("#{delete_column}_changed?")
-                    return if send("#{delete_column}_was").nil?
+                    return unless saved_change_to_logical_delete_column?(delete_column)
+                    return if logical_delete_column_before_last_save(delete_column).nil?
                     return unless send(delete_column).nil?
                     record = send(reflection.name)
                     record.class.increment_counter(cache_column, record.id) unless record.nil?
                   end
+
+                  if ActiveRecord.version < Gem::Version.new('5.1.0')
+                    def saved_change_to_logical_delete_column?(delete_column)
+                      __send__("#{delete_column}_changed?")
+                    end
+
+                    def logical_delete_column_before_last_save(delete_column)
+                      __send__("#{delete_column}_was")
+                    end
+                  else
+                    def saved_change_to_logical_delete_column?(delete_column)
+                      __send__("saved_change_to_#{delete_column}")
+                    end
+
+                    def logical_delete_column_before_last_save(delete_column)
```

Comment:
`attribute_was` seems to be renamed to `attribute_in_database`, is this intended change?
https://github.com/rails/rails/pull/25337/files#diff-60c25306d36b9a651f87d1c4bb381b33be05620f6fc17ec2ddb77b2ba0011681R214
I want some explanation about this 🤔

Reply:
Like I said above, we need to know what was the persisted value before the last save in the after_update callback. This is only possible with `attribute_before_last_save`.

```ruby
require 'rspec'
require 'spec_helper'

puts "# activerecord #{ActiveRecord.version}"
puts

def print_attr_states(at, r)
  ar_5_0 = ActiveRecord.version.to_s < '5.1.0'
  puts <<~END
    # #{at}
    attribute_was              : #{r.attribute_was(:deleted_at).inspect}
    attribute_in_database      : #{ar_5_0 ? '(undef)' : r.attribute_in_database(:deleted_at).inspect}
    attribute_before_last_save : #{ar_5_0 ? '(undef)' : r.attribute_before_last_save(:deleted_at).inspect}

  END
end

# Set non-nil initial value to make it clear that `attribute_before_last_save` always return nil in the before_update callback.
r = Logical.create
r.deleted_at = Time.at(0)
r.save

Logical.before_update { |r| print_attr_states('before_update', r) }
Logical.after_update { |r| print_attr_states('after_update', r) }

r.deleted_at = Time.now
r.save
```

```
% bundle exec ruby -I spec attr_was.rb
# activerecord 5.0.7.2

# before_update
attribute_was              : 1970-01-01 00:00:00 UTC
attribute_in_database      : (undef)
attribute_before_last_save : (undef)

# after_update
attribute_was              : 1970-01-01 00:00:00 UTC
attribute_in_database      : (undef)
attribute_before_last_save : (undef)
```

```
% bundle exec ruby -I spec attr_was.rb
# activerecord 5.1.7

# before_update
attribute_was              : 1970-01-01 00:00:00 UTC
attribute_in_database      : 1970-01-01 00:00:00 UTC
attribute_before_last_save : nil

DEPRECATION WARNING: The behavior of `attribute_was` inside of after callbacks will be changing in the next version of Rails. The new return value will reflect the behavior of calling the method after `save` returned (e.g. the opposite of what it returns now). To maintain the current behavior, use `attribute_before_last_save` instead. (called from print_attr_states at attr_was.rb:11)
DEPRECATION WARNING: The behavior of `attribute_changed?` inside of after callbacks will be changing in the next version of Rails. The new return value will reflect the behavior of calling the method after `save` returned (e.g. the opposite of what it returns now). To maintain the current behavior, use `saved_change_to_attribute?` instead. (called from print_attr_states at attr_was.rb:11)
DEPRECATION WARNING: The behavior of `changed_attributes` inside of after callbacks will be changing in the next version of Rails. The new return value will reflect the behavior of calling the method after `save` returned (e.g. the opposite of what it returns now). To maintain the current behavior, use `saved_changes.transform_values(&:first)` instead. (called from print_attr_states at attr_was.rb:11)
# after_update
attribute_was              : 1970-01-01 00:00:00 UTC
attribute_in_database      : 2021-06-15 10:41:01.792418 UTC
attribute_before_last_save : 1970-01-01 00:00:00 UTC
```

```
% bundle exec ruby -I spec attr_was.rb
# activerecord 5.2.6

# before_update
attribute_was              : 1970-01-01 00:00:00 UTC
attribute_in_database      : 1970-01-01 00:00:00 UTC
attribute_before_last_save : nil

# after_update
attribute_was              : 2021-06-15 10:41:04.842825 UTC
attribute_in_database      : 2021-06-15 10:41:04.842825 UTC
attribute_before_last_save : 1970-01-01 00:00:00 UTC
```